File: projects/order-creation/app/src/main/java/com/robinfood/app/usecases/validatecoupons/IValidateCouponUseCase.java
package com.robinfood.app.usecases.validatecoupons;

import com.robinfood.core.dtos.transactionrequestdto.TransactionRequestDTO;
import com.robinfood.core.enums.TransactionCreationResult;
import lombok.NonNull;

import java.util.concurrent.CompletableFuture;

public interface IValidateCouponUseCase {

    CompletableFuture<TransactionCreationResult> invoke(
            @NonNull String token,
            @NonNull TransactionRequestDTO transactionRequestDTO
    );
}

File: projects/posv2/app/src/main/kotlin/com/robinfood/app/usecases/validatecoupon/IValidateCouponUseCase.kt
package com.robinfood.app.usecases.validatecoupon

import com.robinfood.core.dtos.coupons.CouponValidationRequestDTO
import com.robinfood.core.dtos.coupons.CouponValidationResponseDTO
import com.robinfood.core.enums.Result

/**
 * Use case that validates a coupon
 */
interface IValidateCouponUseCase {

    /**
     * Sends a request to validate a coupon
     * @param token the authentication token
     * @param couponValidationRequestDTO the json body request with the coupon info
     * @return the coupon validation response
     */
    suspend fun invoke(
            token: String,
            couponValidationRequestDTO: CouponValidationRequestDTO
    ): Result<CouponValidationResponseDTO>
}